pkg/fqdn/name_manager.go

```diff
@@ -332,36 +328,37 @@ func (n *NameManager) updateDNSIPs(lookupTime time.Time, updatedDNSIPs map[strin
 			continue
 		}
 
-		addrsToUpsert.Insert(addrs...)
-
 		// record the IPs that were different
-		updatedNames[dnsName] = lookupIPs.IPs
+		updatedNames[dnsName] = addrs
 
-		// accumulate the new selectors affected by new IPs
+		// accumulate the new labels affected by new IPs
 		if len(n.allSelectors) == 0 {
 			log.WithFields(logrus.Fields{
 				"dnsName":   dnsName,
 				"lookupIPs": lookupIPs,
 			}).Debug("FQDN: No selectors registered for updates")
+			continue
 		}
-		for fqdnSel, fqdnRegex := range n.allSelectors {
-			matches := fqdnRegex.MatchString(dnsName)
-			if matches {
-				affectedSelectors.Insert(fqdnSel)
-			}
+
+		// derive labels for this DNS name
+		nameLabels := deriveLabelsForName(dnsName, n.allSelectors)
+		if len(nameLabels) == 0 {
+			// If no selectors care about this name, then skip IPCache updates
+			// for this name.
+			// If any selectors/ are added later, ipcache insertion will happen then.
+			continue
+		}
+
+		updatedMetadata[dnsName] = nameMetadata{
+			addrs:  addrs,
+			labels: nameLabels,
 		}
 	}
 
 	// If new IPs were detected, and these IPs are selected by selectors,
 	// then ensure they have an identity allocated to them via the ipcache.
-	//
-	// If no selectors care about this name, then skip this step. If any selectors
-	// are added later, ipcache insertion will happen then.
-	if len(addrsToUpsert) > 0 && affectedSelectors.Len() > 0 {
-		ipcacheRevision = n.upsertMetadata(addrsToUpsert.UnsortedList())
-	}
-
-	return affectedSelectors, updatedNames, ipcacheRevision
+	ipcacheRevision = n.updateMetadata(updatedMetadata)
```

Comment:
Does it make sense to detect even more no-ops here? I realize this will short-circuit if the name's IPs haven't changed, but what about if the IP's labels haven't changed, just there's a new name.

This might be too much tracking, in which case it's probably not worth it. But it might be worth the bother. This can also be a follow-up PR.

Reply:
If there is a new name, then we ought to still update the IPCache metadata - since the new name needs to be a resource owner of the selector labels as well, even if it shares the same label selectors as the old name.

This is because the old name might be removed from the DNS cache, in which case the prefix selector labels still need to be kept in the IPCache metadata. This is currently achieved by having each name be it's own IPCache metadata resource owner. Thus, if there is a new name, then there is a new resource owner to keep the metadata alive.

We could of course add some tracking here to detect these cases here, but it would basically duplicate the IPCache metadata resource tracking, which I'm not a fan of. I'd rather have IPCache detect the no-op and skip the work (at least it skips the updates to BPF maps already).

Reply:
Fair enough, we can file this away as a possible improvement if necessary. An opportunity to trade memory for CPU.